### ISSUE TITLE
Ajax Load More functionality integration issue

### ISSUE THREAD

**Issue by danielecelsa** [10up/ElasticPress#2023]
Hello,

if in the Search Results page, instead of using default Pagination, I use the Ajax Load More functionality that lots of plugins come with, the results in the first page are correctly handled by ElasticPress, BUT the additional results (after Load More call) are handled by default WP search engine. 

It seem that default WP query is used instead of ES query when Ajax is called.
This does not happen with default pagination (guess because Ajax is not called).

Just as an example, the plugin I am actually using for the Ajax Load more functionality is using this code to handle pagination:

`public function get_document_data(){
 
    global $wp_query;
    

    $id = $this->widget_id;

    $post_id = $this->post_id;
    $theme_id = $this->theme_id;
    $old_query = $wp_query->query_vars;

    $this->query['paged'] = $this->current_page; // we need current(next) page to be loaded
    $this->query['post_status'] = 'publish';

    $wp_query = new \WP_Query($this->query);
    wp_reset_postdata();//this fixes some issues with some get_the_ID users.
    if (is_archive()){
      $post_id = $theme_id;
    }
 
		$document = \Elementor\Plugin::$instance->documents->get_doc_for_frontend( $post_id );
		$theme_document = \Elementor\Plugin::$instance->documents->get_doc_for_frontend( $theme_id );
   
    $data[] = $this->get_element_data($id,$theme_document->get_elements_data());

		// Change the current post, so widgets can use documents->get_current.
		\Elementor\Plugin::$instance->documents->switch_to_document( $document );

    ob_start();
        $document->print_elements_with_wrapper( $data );
    $content = ob_get_clean();
    echo $this->clean_response($content,$id);

    \Elementor\Plugin::$instance->documents->restore_document();
    $wp_query->query_vars = $query_vars;

    die;
  }`

Any idea how to make this work?

Thanks,
Daniele

**Comment by brandwaffle** [comment]
@danielecelsa at a high level, ElasticPress works by "taking over" certain queries. This includes the "s" parameter in a query, which indicates a search. Depending on how the Ajax plugin you're using works (they're all different), the author might be creating a special non-search query, or they could be even issuing a direct SQL query, which means ElasticPress can't "see" the query to take it over.

In order to make any query use ElasticPress, you just need to add a parameter of **ep_integrate** and set it to **true**. This can be done with a filter like **pre_get_posts** if you can't directly add it to the WP_Query call (likely the case with a custom plugin).

**Comment by brandwaffle** [comment]
@danielecelsa I hear you're working on this with our support team as well, in which case I'll go ahead and close the ticket in 3 days unless you want me to keep this open. Thanks!

**Comment by danielecelsa** [comment]
Hello @brandwaffle , please do not close this ticket yet.
I think it is better to talk about this here in public, maybe someone could find it helpful. Also I will not go on with the conversation with support team, so let's continue here if it is possible :) 

About the issue, I did what you suggested and I got some results, I guess the issue now is about 'pagination'.

I tried this code as per your suggestion:
`  

    $this->query['paged']= 1; // did not work with $this->query['paged'] = $this->current_page;
    $this->query['post_status'] = 'publish';
    $this->query['ep_integrate'] = 'true';  // EP support suggestion

    $wp_query = new \WP_Query($this->query);

`

and when I click on Load More button to load the page 2, in this new page I see the same results of the first page.
Means EP query is called, but it shows its first results, as if it was called for the first time!

So could you suggest how can I do to make the query show its results 'related' to the page shown? 

Thanks in advance!
Daniele

**Comment by danielecelsa** [comment]
Hello @brandwaffle,

did you by chance have the opportunity to deepen my request?

Thanks a lot,
Daniele

**Comment by danielecelsa** [comment]
@brandwaffle I have an **update**

Actually pagination works, and I got which is the real issue:
The problem is ONLY in the **LAST page**.

The Ajax Load More plugin handle well the query through ElasticPress and also pagination, it only fails in the last page because it shows the last results of ElasticPress query **PLUS** some results of WP query, if the number of post per page is more than the last results of EP query.

Little **example** to easily explain:
EP query gives 11 results
Number of post per page = 5
So I search and I see the first 5 results of EP, then I click Load More and I see other 5 results of EP,  then I click again Load More and i should see the 11th result, only that.
**BUT**... I see that result PLUS 4 more, and these 4 come by the default WP query I guess.

This does NOT happen with the default WP pagination. 
I mean, if instead of 'Load More' I use 'Numbers' or 'Next/Previous' in pagination settings of Posts widget, in the last page I will see only the 11th result.

### So I would like to ask you:
do you know what part of code/parameter of EP deals with this feature of the last results in pagination? 
I guess somewhere it must be specified.

And I guess that the Load More plugin in someway overwrites this parameter (because without plugin, in normal pagination, this does not happen), but I really don't know how to identify it.

Thank you in advance,
Daniele

**Comment by felipeelia** [comment]
Hi @danielecelsa ,

ElasticPress should not be changing anything on the last page. I have a suggestion to help you to debug what is going on.

1. Enable WP_DEBUG and WP_DEBUG_LOG constants in your wp-config.php file. Set WP_DEBUG_DISPLAY as false as well.
2. Inside your `get_document_data()` function, right before you call `$wp_query = new \WP_Query($this->query);`, insert `error_log( var_export( $this->query, true ) );`
3. Try to load the pages

It will store the parameters in a `debug.log` file inside your `wp-content` folder. Hopefully, that can give us an idea of what WordPress (and ElasticPress) will be looking for. If you'd like, you can also place a `error_log( var_export( $wp_query->posts, true ) );` there too, to check what is coming back.

Thanks

**Comment by danielecelsa** [comment]
Hello @felipeelia,

first, really thank you very much for you kind help!

Then, I tried your suggestion and here my results:
- I attach _debug_log_github.txt_ that is an extract of the debug.log file, related to the test I just tried.
-  I attach also what I have in my wp-config.php file. I realized that I already had some of your suggestion in that file, so I just added `define( 'WP_DEBUG_DISPLAY', false );` as you can see (screenshot _wp_config_).
- Finally I attach the last part of the function `get_document_data()`,  where I inserted what you suggested (screenshot _plugin_function_).

In this last screenshot, I highlighted the row 442 where the debug.log file shows there is an error. I do not know if it is related to the issue (I just asked to the owner of the plugin).

To help you understanding:
- I searched for _sushi_ term
- I have set 5 post per page
- The issue is in page 4, the last one, where I see the last of the 16 results that EP should give me but I see also other 4 results coming, I guess, from WP default query.
(The first 3 pages correctly show, 5 by 5, the first 15 results that EP gives) 

I really hope you can find it helpful in the investigation of the issue, I tried to find the problem but unfortunately I could not end up with any solution..

Thanks in advance for your kind help,
Daniele

<img width="794" alt="plugin_function" src="https://user-images.githubusercontent.com/77167372/106332297-cc708d00-6286-11eb-9d76-1f3ab308daca.png">
<img width="564" alt="wp_config" src="https://user-images.githubusercontent.com/77167372/106332303-cf6b7d80-6286-11eb-9967-580050b7b3ef.png">
[debug_log_github.txt](https://github.com/10up/ElasticPress/files/5896528/debug_log_github.txt)

**Comment by felipeelia** [comment]
Hi @danielecelsa ,

The `error_log( var_export( $wp_query->posts, true ) );` should be placed _after_ the `$wp_query = new \WP_Query...` line. Can you please make that change, try again and share the results? Thanks

**Comment by danielecelsa** [comment]
Hi @felipeelia,

ok sorry, it was a misunderstanding.

Here new debug.log file, hope this could help:
[debug_log_github_2.txt](https://github.com/10up/ElasticPress/files/5914205/debug_log_github_2.txt)

Thank you in advance for your kind help!
Regards,
Daniele

**Comment by danielecelsa** [comment]
Hello @felipeelia 

did you maybe have the chance to have a look to my last message/log file?
Or maybe it was incomplete and you could not make any investigation?

Kindly let me know if you need further info to help me :)

Thank you in advance,
Daniele

**Comment by felipeelia** [comment]
Hi @danielecelsa ,

Looking at the logs, it looks like ElasticPress may be looking for "sushi" in other fields or taxonomies. Did you already try to disable everything in the Weighting Dashboard and leaving only "Title" enabled?

**Comment by danielecelsa** [comment]
Hi @felipeelia ,

EP looking for 'sushi' in other fields or taxonomy is exactly what I implemented and what I'd like to have.
And it is working fine in all the pages, except the last one.

The problem is that, in the last page, the results are coming like they are searched with the default 'WP search', so maybe only in the 'Title' or however it works.

Hope to have been clear, kindly let me know if you need further details.
And thanks again for your time! :)
Daniele

**Comment by danielecelsa** [comment]
Hi @felipeelia ,

did you maybe had the chance to have a look at my last comment? 

Feel free to let me know if I have not been clear, so I will try to explain better.

Thanks!
Daniele

**Comment by felipeelia** [comment]
Hi @danielecelsa ,

We'll need to give a look at the queries sent to the Elasticsearch server to figure out why it is returning those results. Do you have any kind of log that contains it? Could you please share it with us? 

The regular WP search will use just the product title and content. ElasticPress, on the other hand, will use those fields and some taxonomies too. Did you try disabling everything in the Weighting Dashboard and leaving just "Title" enabled?

**Comment by felipeelia** [comment]
As we didn’t have any movement in this issue for a while now, we are now closing it. If you’re still facing the issue, please let us know. Otherwise, if you’ve managed to get an answer to your problem, we’d highly appreciate it if you could get back to us here and leave the solution in a comment. Thanks!

**Comment by danielecelsa** [comment]
Hello again @felipeelia and all,

sorry I have been very busy and could not answer, so you correctly closed the case. 
Wonder if you can open it again and help me out.

Besides the test you asked to make, I would like to ask you a basic thing that maybe is what I am missing:
How/where exactly should I set **ep_integrate = true** ?

I did it on the plugin itself (the one of Load More), but maybe it is not working.

You suggested to use _pre_get_posts_, but actually I don't know how.
I tried to write this in _function.php_:

`
function ep_integrate_function( $query ) {
  if ( !is_admin() && $query->is_main_query() ) {
      $query->set('ep_integrate', true);
    
  }
}
add_action( 'pre_get_posts', 'ep_integrate_function' );
`

But it did not work, actually I receive a _Page Not Found_ when I try to reach my website, so I deleted it.

I am asking this because I noted that in the Elementor template for Search Result page, if I use a 'preview' word to see which results will be shown, those results are **not** the ones shown by ES in a 'real' search!
And Load More button actually load THOSE results and not the ones in next pages of ES results.

So, could you kindly help me make all the queries to be integrated with ES, so that maybe also Load More will 'see' it?

Hope to have been clear, kindly let me know if not.

Thanks in advance!
Daniele

**Comment by danielecelsa** [comment]
Hello,

did you have the chance to have a look to my comment?

I am writing you also to ask if you need debug.log file, which configuration you would need me to set? 
- with _ep_integrate=true_ in the plugin
- with _pre_get_post_ set in _function.php_ (if yes, could you tell me if how I wrote it is correct?)
- both

Consider that, with or without '_ep_integrate =true_' in the plugin, I see this in the debug.log:
<img width="264" alt="Schermata 2021-04-14 alle 20 41 04" src="https://user-images.githubusercontent.com/77167372/114762247-d3754a80-9d61-11eb-9391-3886b6637fd3.png">

That part in debug.log is shown thanks to command:
`error_log( var_export( $wp_query->posts, true ) );` 

Furthermore, if I put '_ep_integrate =true_' in the plugin, I also see this in the debug.log
<img width="226" alt="Schermata 2021-04-14 alle 20 44 22" src="https://user-images.githubusercontent.com/77167372/114762903-847be500-9d62-11eb-81ac-80e001cb1810.png">

in the part shown thanks to command:
`error_log( var_export( $this->query, true ) );`


Anyway nothing change, the results shown are the same with or without '_ep_integrate =true_'

The way I insert '_ep_integrate =true_' in the file _ajax-pagination.php_ is:
`$this->query['ep_integrate'] = true;`

Thanks in advance for your kind help,
Daniele

**Comment by pinguluk** [comment]
Hey @danielecelsa, are you using [Elementor Custom Skin](https://wordpress.org/plugins/ele-custom-skin/) for the load more?
I ran into the same or similar issue as well these days.

I use [JetSmartFilters](https://crocoblock.com/plugins/jetsmartfilters/) to filter some posts (Elementor Posts Widget) on a page and it seems that after I filter the posts, it renders the posts, but just based on its query.

I tested out the load more from Elementor Custom Skin and it seems that when you click the **Load More** button, it sends an ajax request, to retrieve the posts.
In this ajax request, it sends these parameters:

`action: ecsload`

```
query: {
   "page":0,
   "pagename":"news-and-events",
   "error":"",
   "m":"",
   "p":0,
   "post_parent":"",
   "subpost":"",
   "subpost_id":"",
   "attachment":"",
   "attachment_id":0,
   "name":"news-and-events",
   "page_id":0,
   "second":"",
   "minute":"",
   "hour":"",
   "day":0,
   "monthnum":0,
   "year":0,
   "w":0,
   "category_name":"",
   "tag":"",
   "cat":"",
   "tag_id":"",
   "author":"",
   "author_name":"",
   "feed":"",
   "tb":"",
   "paged":0,
   "meta_key":"",
   "meta_value":"",
   "preview":"",
   "s":"",
   "sentence":"",
   "title":"",
   "fields":"",
   "menu_order":"",
   "embed":"",
   "category__in":[
      
   ],
   "category__not_in":[
      
   ],
   "category__and":[
      
   ],
   "post__in":[
      
   ],
   "post__not_in":[
      
   ],
   "post_name__in":[
      
   ],
   "tag__in":[
      
   ],
   "tag__not_in":[
      
   ],
   "tag__and":[
      
   ],
   "tag_slug__in":[
      
   ],
   "tag_slug__and":[
      
   ],
   "post_parent__in":[
      
   ],
   "post_parent__not_in":[
      
   ],
   "author__in":[
      
   ],
   "author__not_in":[
      
   ],
   "post_type":[
      "post",
      "page",
      "e-landing-page"
   ],
   "ignore_sticky_posts":false,
   "suppress_filters":false,
   "cache_results":true,
   "update_post_term_cache":true,
   "lazy_load_term_meta":true,
   "update_post_meta_cache":true,
   "posts_per_page":10,
   "nopaging":false,
   "comments_per_page":"50",
   "no_found_rows":false,
   "order":"DESC"
}
```

```
ecs_ajax_settings: {
   "current_page":1,
   "max_num_pages":99,
   "load_method":"loadmore",
   "widget_id":"6c79da20",
   "post_id":"4120",
   "theme_id":"5499"
}
```

I tried to change the **query** parameter, but it seems that it doesn't matter at all! It ignores it completely!
After more digging, I found out that the **widget_id**, **post_id** and **theme_id** matter.
**widget_id** is the id from the Elementor Post Widget
**post_id** is the id of the page where the post widget is
**theme_id** is the id of the template that you are using to display the custom template for the post widget

Then I've looked into the `includes/ajax-pagination.php` file to try to find the problem, but could not pinpoint the problem exactly. It seems that it just ignores the **query** parameter from the ajax request and it doesn't affect the final query. It will query based of the settings of the elementor post widget (e.g. if you edit the widget and set the query to include the terms 'sushi', then the load more function will only load posts that include the term (taxonomy) 'sushi')

I've edited the file and added these lines:

```
$this->customFilterQuery = json_decode(
    stripslashes($_POST['custom_filter_term_ids']),
    true
);
```

before these lines:

```
if ($this->current_page > $args['max_num_pages']) {
    return;
}
```

and

``` 
if (isset($this->customFilterQuery) && $this->customFilterQuery["term_ids"] != "") {
    $data[0]["settings"]["posts_include"] = ["terms"];
    $data[0]["settings"]["posts_include_term_ids"] = explode(
        ",",
        $this->customFilterQuery["term_ids"]
    );
}
```

before these lines:

```
// Change the current post, so widgets can use `documents->get_current`.
\Elementor\Plugin::$instance->documents->switch_to_document($document);
 ```
 
 
I've also edited the `assets/js/ecs_ajax_pagination.js` file and added a parameter called `custom_filter_term_ids`, that has a list of terms ids (from the selectors from the filter). Basically it will send the ids for the terms that you want, to the backend function and will "override" the post widget settings. It's a hacky workaround, but it works.

I'm bit in a hurry, so I will update my answer later for more info.
Long story short: there's a bug (?) in the elementor custom skin plugin, that ignores the query parameter from the ajax and there's an incompability with any plugin that affects the rendering (elasticpress, jetsmartfilters etc) and you need to come with a workaround, by passing the terms or anything you want to filter, in the ajax and use them to override the post widget settings, because the WP_Query also doesn't work (?).

**Comment by danielecelsa** [comment]
Hello @pinguluk 

WOW thank you for sharing!

I am in hurry too, so I will check better later all your really interesting material :)

Anyway Yes, I am using Ele Custom Skin and there is for sure a bug to fix, but unfortunately the creator of this plugin is always busy..

For now, I just want to add that I was wondering if the issue was about these lines in ajax-pagination.php:
` $old_query = $wp_query->query_vars;`

and 

`$wp_query->query_vars = $query_vars;`

The second one throws an error if you check the file debug.log. To get it, you need too add the following lines that @felipeelia kindly suggested at the beginning of this thread:
`error_log( var_export( $this->query, true ) );`
or
`error_log( var_export( $wp_query->posts, true ) );`

I don't remember which one is actually printing the log I meant.

The error should be something like '$query_vars is not defined'

Soo maybe the right line should be:
`$wp_query->query_vars = $old_query;`

otherwise $old_query is not used in the function.

So I was wondering if maybe the creator wanted to use that approach but it made an error in the syntax.
Maybe the 'old' query_vars, that should be the vars of the original first query, should be somehow used in the creation of the new query.

Something like:
`$this->query['query_vars'] = $old_query;`

but I still did not have time to try, and anyway I am sure the syntax is incorrect :)

Anyway I will try to use your HOTFIX in order to workaround also my situation with ElasticSearch, thank you very much for sharing!
I will let you know :)

Daniele

**Comment by pinguluk** [comment]
I will look again into that thing with `$wp_query->query_vars` later, maybe I will find something interesting. Also, I think you'll need to adapt the hotfix, because I was querying by terms (tag taxonomy) and you are querying by search (which is querying the posts if they include the searched string) and I don't know if you can set somewhere in the post widget settings to filter by that (try to var_dump the `$data` variable and look what else can you modify)

**Comment by Junior774** [comment]
Hi @pinguluk 

I think I'm having the same problem and I'm trying to implement your solution, where do I set the ids terms? and where I can see the ids of my terms, I'm using advanced query options dinamyc related posts.

**Comment by pinguluk** [comment]
> Hi @pinguluk
> 
> I think I'm having the same problem and I'm trying to implement your solution, where do I set the ids terms? and where I can see the ids of my terms, I'm using advanced query options dinamyc related posts.

From what I remember and read from my comment, on **initial load** and every **load next page**, I call the custom **getFilterIds** function, which retrieves the ids of the posts and set them in the **data**, for the ajax request:

```javascript
data = {
        'action': 'ecsload',
        'query': ecs_ajax_params.posts,
        'ecs_ajax_settings': settings,
        // HOTFIX: added custom filter term ids into ajax, so we can query them from the backend, because the "query" parameter doesn't do anything at ALL
        // Because there's a bug? and the load more button loads posts based on the Elementor Widget settings
        'custom_filter_term_ids': '{"term_ids":"' + custom_filter_term_ids.join(',') + '"}'
    };
```

then in the php file, I retrieve and assign the data to a new variable

```php
$this->customFilterQuery = json_decode(
    stripslashes($_POST['custom_filter_term_ids']),
    true
);
```

and then override the data, to be queried.

```php
if (isset($this->customFilterQuery) && $this->customFilterQuery["term_ids"] != "") {
    $data[0]["settings"]["posts_include"] = ["terms"];
    $data[0]["settings"]["posts_include_term_ids"] = explode(
        ",",
        $this->customFilterQuery["term_ids"]
    );
}
```

which basically sets the **posts_include_term_ids** to the specified **term ids**.

Just follow the code and try to figure it out

**Comment by raulillana** [comment]
Hey @danielecelsa! Did you solve this with the patch from @pinguluk?

Actually using searchandfilter query and messing around this function too...

If any of you guys has a patch, please, make a PR in the plugin github so others like me can benefit from it.

https://github.com/dudaster/ele-custom-skin/

Thanks in advance!
